### PR TITLE
Improve AArch64 depthwise convolution through smlal/smlal2 intrinsic

### PR DESCRIPTION
**Pull request by giuseros**
- Added an intrinsic to load a single int16x8 vector and produce two
  int32x4 output vectors through smlal/smlal2 instructions

- Changed the NHWC depthwise schedule to accomodate the aforementioned
  intrinsic

Change-Id: I347c3bf98fa8dd87057304dcda0d78e558424c57

### CODE REVIEW

**Comment by giuseros** [comment]
cc @u99127 @anijain2305

**Comment by FrozenGene** [comment]
I like this change. May I ask two questions?

1. could we make `smlal` / `smlal2` work on convolution too?
2. If we don't use tensorize, do we have method to handle this? For example, add optimization of custom LLVM pass or low detail analyze on tir or llvm ir?

**Comment by giuseros** [comment]
Hi @FrozenGene 
As for your questions:
1) Yes we can, but there is a issue. It's very hard to have the two quantized strategies (int8 + smull/sadalp vs int16 + smlal/smlal2) available to the auto-tuner at the same time (you already saw [this post](https://discuss.tvm.apache.org/t/quantized-models-and-legalization-pass/8253/3), I will reply there to your comment). In the case of depthwise, we only have the int16 strategy, so the issue does not arise. 
2) Nice you asked, because I am trying to prototype  the idea of a "tir pass" where this  tensorization (and possibly every arm tensorization in `tensor_intrin.py`) happens. My hope is to make it work more or less like `vectorize`. One of the main reasons to have this is to enable Ansor support

**Comment by FrozenGene** [comment]
> Hi @FrozenGene
> Thanks for your reply!
> 
> 1. Yes we can, but there is a issue. It's very hard to have the two quantized strategies (int8 + smull/sadalp vs int16 + smlal/smlal2) available to the auto-tuner at the same time (you already saw [this post](https://discuss.tvm.apache.org/t/quantized-models-and-legalization-pass/8253/3), I will reply there to your comment). In the case of depthwise, we only have the int16 strategy, so the issue does not arise.
> 2. Nice you asked, because I am trying to prototype  the idea of a "tir pass" where this  tensorization (and possibly every arm tensorization in `tensor_intrin.py`) happens. My hope is to make it work more or less like `vectorize`. One of the main reasons to have this is to enable Ansor support

For ansor, we could make ansor support tensorize (like TensorCore we need tensorize too). However, if we could done it in the llvm / tir, we will make ansor support it easily. If we could lift it as generic pass, I think it will bring benifit to other hardware platform too.

File: python/tvm/topi/arm_cpu/tensor_intrin.py [hunk 1]
```diff
@@ -879,6 +879,96 @@ def _instr(index):
...
+            ib = tvm.tir.ir_builder.create()
+            if index == 1:
+                ib.emit(outs[0].vstore(0, tvm.tir.const(0, "int32x8")))
+                return ib.get()
+
+            vec_a = ins[0].vload([0], "int16x8")
+            vec_b = ins[1].vload([0, 0], "int16x8")
+            inst = "llvm.aarch64.neon.smull"
+
+            # Lower part of the vector
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/tensor_intrin.py:935]: Higher part of the vector.

File: python/tvm/topi/arm_cpu/tensor_intrin.py [hunk 2]
```diff
@@ -879,6 +879,96 @@ def _instr(index):
...
+            vec_c_l = outs[0].vload([0], "int32x4")
+            vec_a_l = tvm.tir.call_intrin("int16x4", "tir.vectorlow", vec_a)
+            vec_b_l = tvm.tir.call_intrin("int16x4", "tir.vectorlow", vec_b)
+            vmull_l = tvm.tir.call_llvm_pure_intrin(
+                "int32x4", inst, tvm.tir.const(2, "uint32"), vec_a_l, vec_b_l
+            )
+            vec_out_l = vec_c_l + vmull_l
+
+            # Combine higher and lower part in a single int32x8 vector to store
+            # (this will require two different STR instructions, since the
```
Comment (mbaret) [inline python/tvm/topi/arm_cpu/tensor_intrin.py:954]: close brackets, STR -> STORE (for those of us who play with strings a lot :) )

**Review by mbaret — CHANGES_REQUESTED** [review]

**Review by mbaret — APPROVED** [review]
LGTM now. I agree with @FrozenGene that this would be best implemented nearer the LLVM codegen for reusability, but I think this is a good start and demonstrates a worthwhile benefit to performance.

**Comment by giuseros** [comment]
Hi @mbaret , thanks for the review! 

@FrozenGene , any update on this?

**Review by FrozenGene — APPROVED** [review]

**Comment by FrozenGene** [comment]
@giuseros @mbaret Thanks, merged now

**Pull request merged**